apis/mutations/unversioned/externaldata.go

```diff
@@ -0,0 +1,167 @@
+// +kubebuilder:validation:Enum=ValueAtLocation;Username
+type DataSource string
+
+const (
+	// ValueAtLocation indicates that the value at spec.location of the mutation
+	// spec will be extracted to the external data provider as the data source.
+	DataSourceValueAtLocation DataSource = "ValueAtLocation"
+
+	// Username indicates that the username of the admission request will
+	// be extracted to the external data provider as the data source.
+	DataSourceUsername DataSource = "Username"
+)
+
+// FailurePolicy is the type of the failure policy to use for the external data.
+// +kubebuilder:validation:Enum=UseDefault;Ignore;Fail
+type FailurePolicy string
+
+const (
+	// UseDefault indicates that the default value of the external
+	// data provider will be used.
+	FailurePolicyUseDefault FailurePolicy = "UseDefault"
+
+	// Ignore indicates that the mutation will be ignored if the external
+	// data provider fails.
+	FailurePolicyIgnore FailurePolicy = "Ignore"
+
+	// Fail indicates that the mutation will be failed if the external
+	// data provider fails.
+	FailurePolicyFail FailurePolicy = "Fail"
+)
+
+// ExternalData describes the external data source to use for the mutation.
+type ExternalData struct {
+	// Provider is the name of the external data provider.
+	// +kubebuilder:validation:Required
+	Provider string `json:"provider,omitempty"`
+
+	// DataSource specifies where to extract the data that will be sent
+	// to the external data provider as parameters.
+	// +kubebuilder:default="ValueAtLocation"
+	DataSource DataSource `json:"dataSource,omitempty"`
+
+	// FailurePolicy specifies the policy to apply when the external data
+	// provider returns an error.
+	// +kubebuilder:default="Fail"
+	FailurePolicy FailurePolicy `json:"failurePolicy,omitempty"`
+
+	// Default specifies the default value to use when the external data
+	// provider returns an error and the failure policy is set to "UseDefault".
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:validation:XPreserveUnknownFields
+	Default *types.Anything `json:"default,omitempty"`
+}
+
+func (e *ExternalData) Validate() error {
+	if !*externaldata.ExternalDataEnabled {
+		return ErrExternalDataFeatureFlagNotEnabled
+	}
+	if e.FailurePolicy == FailurePolicyUseDefault && e.Default == nil {
+		return ErrExternalDataFeatureNoDefault
+	}
+
+	return nil
+}
+
+// GetValue sends a request to the external data provider to get the value.
+// It returns a map of the outbound data and its response as key-value pairs.
+func (e *ExternalData) GetValue(param *types.GetValueParameter) (map[types.ProviderOutboundData]interface{}, error) {
+	sendRequest := types.DefaultSendRequestToExternalDataProvider
+	if param.SendRequestToExternalDataProvider != nil {
+		sendRequest = param.SendRequestToExternalDataProvider
+	}
+
+	provider, err := param.ProviderCache.Get(e.Provider)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get external data provider %s from cache: %v", e.Provider, err)
```

Comment:
Here and throughout, consider wrapping the inner error via the `%w` formatting verb.
```suggestion
		return nil, fmt.Errorf("failed to get external data provider %s from cache: %w", e.Provider, err)
```